marlon/baseline_models/env_wrappers/defend_wrapper.py

```diff
@@ -0,0 +1,259 @@
+            node_id = get_node_from_action(node_from_action)
+            node_info = get_node_info(node_id)
+            return node_id, node_info
+
+        def get_node_from_action(node_from_action: int):
+            """Gets the node id from an action"""
+            return list(self.cyber_env.environment.network.nodes)[node_from_action]
+
+        def get_node_info(node_id: model.NodeID):
+            """Given a node ID, find the corresponding node info"""
+            return self.cyber_env.environment.get_node(node_id)
+
+
+        def node_exists(node_id: model.NodeID):
+            """Determines if a node exists in the network"""
+            if node_id in list(self.cyber_env.environment.network.nodes):
+                return True
+            return False
+
+        def node_running(node_info: model.NodeInfo):
+            """Determines if a node is currently running"""
+            if node_info.status == model.MachineStatus.Running:
+                return True
+            return False
+
+        def node_exists_and_running(node_from_action: int):
+            """Determines if a node exists in the network, and if so if it is running."""
+            node_id, node_info = get_node_and_info(node_from_action)
+            return (node_exists(node_id) and node_running(node_info))
+
+        def is_reimagable(node_info: model.NodeInfo):
+            """Checks if a given node is reimagable"""
+            if node_info.reimagable:
+                return True
+            return False
+
+        def firewall_rule_exists(node_info: model.NodeInfo, port_from_action: int, incoming :bool):
+            """Checks a node to see if a given firewall rule exists on it."""
+            firewall_list = []
+            if incoming:
+                for rule in node_info.firewall.incoming:
+                    firewall_list.append(rule.port)
+            else:
+                for rule in node_info.firewall.incoming:
+                    firewall_list.append(rule.port)
+
+            if self.firewall_rule_list[port_from_action] not in firewall_list:
+                return False
+            return True
+
+        def service_exists(node_info: model.NodeInfo, service_from_action: int):
+            """Checks if a service exists on a node"""
+            # TODO IS THIS GOOD? This is just checking if its out of bounds...
+            # The names could be in a different order, so action 1 could be one service on one node and a differnet on another...
+            if service_from_action >= len(node_info.services):
+                return False
+            return True
+
+        if action[0] == 0:
+            # REIMAGE
+            _, node_info = get_node_and_info(action[1])
+            # If the node does not exist, or if the node is not currently running, this action is invalid.
+            return node_exists_and_running(action[1]) and is_reimagable(node_info)
+
+        elif action[0] == 1:
```

Comment:
Minor

```suggestion
        action_number = action[0]
        if action_number == 0:
            # REIMAGE
            _, node_info = get_node_and_info(action[1])
            # If the node does not exist, or if the node is not currently running, this action is invalid.
            return node_exists_and_running(action[1]) and is_reimagable(node_info)

        elif action_number == 1:
```

Method is very clean otherwise, nice local functions and comments 👍